tools/wptrunner/wptrunner/browsers/chrome_android.py

```diff
@@ -68,21 +69,72 @@ def env_options():
     # allow the use of host-resolver-rules in lieu of modifying /etc/hosts file
     return {"server_host": "127.0.0.1"}
 
+class LogcatRunner(object):
+    def __init__(self, logger, browser, remote_queue):
+        self.logger = logger
+        self.browser = browser
+        self.remote_queue = remote_queue
+
+    def start(self):
+        try:
+            self._run()
+        except KeyboardInterrupt:
+            self.stop()
+
+    def _run(self):
+        try:
+            self.browser.clear_log()
+            self._cmd = self.browser.logcat_cmd()
+            self._proc = mozprocess.ProcessHandler(
+                self._cmd,
+                processOutputLine=self.on_output,
+                storeOutput=False)
+            self._proc.run()
+        except (OSError, subprocess.CalledProcessError):
+            self.logger.error("Failed to start adb logcat")
+
+    def _send_message(self, command, *args):
+        self.remote_queue.put((command, args))
 
-class ChromeAndroidBrowser(Browser):
-    """Chrome is backed by chromedriver, which is supplied through
-    ``wptrunner.webdriver.ChromeDriverServer``.
-    """
+    def stop(self, force=False):
+        if self.is_alive():
+            kill_result = self._proc.kill()
+            if force and kill_result != 0:
+                self._proc.kill(9)
 
-    def __init__(self, logger, package_name, webdriver_binary="chromedriver",
+    def is_alive(self):
+        return hasattr(self._proc, "proc") and self._proc.poll() is None
+
+    def on_output(self, line):
+        try:
+            data = {
+                "process": "LOGCAT",
+                "command": "logcat",
+                "data": line
+            }
+            self._send_message("log", "process_output", data)
+        except:
+            pass
+
+class AndroidBrowser(Browser):
```

Comment:
I'm not sure I love the name `AndroidBrowser`, because there are many browsers on Android (Chrome, Firefox, Edge, Brave, ...) and this code doesn't support anything outside of Chrome. 

What is the purpose of the `ChromeAndroidBrowser` subclass? All it does is override `setup_adb_reverse`, which seems to have the same override in all three subclasses. Couldn't we just move `setup_adb_reverse` to the super-class (`AndroidBrowser`), delete the current `ChromeAndroidBrowser` and rename `AndroidBrowser --> ChromeAndroidBrowser`.

Reply:
Weizhong and I discussed this, and there is a medium bit of refactoring to do to simplify all this further. I don't want to block this PR, so I agreed that we can do the cleanup in a follow-up PR, and for now we will just rename:

`AndroidBrowser --> ChromeAndroidBrowserBase`.

Reply:
Hi Stephen
Thanks for the review. I have pushed a new commit. Can you pls review again?
Thanks!